### PR TITLE
Catalog2lucene - License Url update for license expression/file

### PR DESCRIPTION
**Pull request by zhhyu**
For catalog2lucene pipeline, when the package contains license expression/file, the license Url should be updated as expected: "~/packages/id/version/license". 

Fix: https://github.com/NuGet/NuGetGallery/issues/6526

### CODE REVIEW

File: src/NuGet.Indexing/Extraction/CatalogPackageMetadataExtraction.cs [hunk 1]
```diff
@@ -55,18 +57,52 @@ private class Extractor
...
 
                 return _metadata;
             }
 
+            private void AddLicenseUrl()
+            {
+                if (_galleryBaseAddress != null && _catalog[MetadataConstants.LicenseExpressionPropertyName] != null)
+                {
+                    _metadata[MetadataConstants.LicenseUrlPropertyName] = GetOverWriteLicenseUrl();
+                } else if (_galleryBaseAddress != null && _catalog[MetadataConstants.LicenseFilePropertyName] != null)
```
Comment (joelverhagen) [inline src/NuGet.Indexing/Extraction/CatalogPackageMetadataExtraction.cs]: nit: `else if` should be on a different line that the curlies

File: src/NuGet.Indexing/Extraction/CatalogPackageMetadataExtraction.cs [hunk 2]
```diff
@@ -55,18 +57,52 @@ private class Extractor
...
+                AddLicenseUrl();
                 AddFlattenedDependencies();
                 AddSupportedFrameworks();
 
                 return _metadata;
             }
 
+            private void AddLicenseUrl()
+            {
+                if (_galleryBaseAddress != null && _catalog[MetadataConstants.LicenseExpressionPropertyName] != null)
```
Comment (skofman1) [inline src/NuGet.Indexing/Extraction/CatalogPackageMetadataExtraction.cs]: This if statement can be simplified:
  if (_galleryBaseAddress != null && 
(_catalog[MetadataConstants.LicenseExpressionPropertyName] != null || _catalog[MetadataConstants.LicenseFilePropertyName] != null))
{
_metadata[MetadataConstants.LicenseUrlPropertyName] = GetOverWriteLicenseUrl();
}
else
{
AddString(MetadataConstants.LicenseUrlPropertyName);
}

File: src/NuGet.Indexing/Extraction/CatalogPackageMetadataExtraction.cs [hunk 3]
```diff
@@ -55,18 +57,52 @@ private class Extractor
...
+                } else if (_galleryBaseAddress != null && _catalog[MetadataConstants.LicenseFilePropertyName] != null)
+                {
+                    _metadata[MetadataConstants.LicenseUrlPropertyName] = GetOverWriteLicenseUrl();
+                } else
+                {
+                    AddString(MetadataConstants.LicenseUrlPropertyName);
+                }
+            }
+
+            private string GetOverWriteLicenseUrl()
```
Comment (skofman1) [inline src/NuGet.Indexing/Extraction/CatalogPackageMetadataExtraction.cs]: We should have a similar method in Ctalog2Registration. Can we share the code?

File: src/V3PerPackage/PerBatchProcessor.cs [hunk 4]
```diff
@@ -282,6 +282,7 @@ private async Task ExecuteCatalog2LuceneAsync(PerBatchContext context, Uri catal
                     commitEachBatch,
                     commitTimeout,
                     baseAddress,
+                    context.Global.GalleryBaseAddress.AbsoluteUri,
```
Comment (agr) [inline src/V3PerPackage/PerBatchProcessor.cs]: >GalleryBaseAddress [](start = 35, length = 18)

Can `GalleryBaseAddress` be null here?

File: src/Ng/SearchIndexFromCatalogCollector.cs [hunk 5]
```diff
@@ -38,6 +39,7 @@ public class SearchIndexFromCatalogCollector : CommitCollector
             bool commitEachBatch,
             TimeSpan? commitTimeout,
             string baseAddress,
+            string galleryBaseAddress,
             ITelemetryService telemetryService,
```
Comment (agr) [inline src/Ng/SearchIndexFromCatalogCollector.cs]: This should probably by a `Uri` object instead of a `string`.

File: src/NuGet.Indexing/Extraction/CatalogPackageMetadataExtraction.cs (hunk 3 above)
Comment (zhhyu) [reply src/NuGet.Indexing/Extraction/CatalogPackageMetadataExtraction.cs]: We have a  sharing license helper to get license Url now.

File: src/V3PerPackage/PerBatchProcessor.cs (hunk 4 above)
Comment (zhhyu) [reply src/V3PerPackage/PerBatchProcessor.cs]: Use Uri rather than string here.

File: src/Ng/SearchIndexFromCatalogCollector.cs (hunk 5 above)
Comment (zhhyu) [reply src/Ng/SearchIndexFromCatalogCollector.cs]: Yes, Uri is better than string.

File: src/NuGet.Indexing/Extraction/CatalogPackageMetadataExtraction.cs [hunk 6]
```diff
@@ -55,18 +57,52 @@ private class Extractor
...
+            {
+                var uriBuilder = new UriBuilder(_galleryBaseAddress);
+                var packageId = JTokenToString(_catalog[MetadataConstants.IdPropertyName]);
+                var packageVersion = JTokenToString(_catalog[MetadataConstants.NormalizedVersionPropertyName]);
+
+                if (packageId == null)
```
[the remark this answers is not in the document]
Comment (zhhyu) [reply src/NuGet.Indexing/Extraction/CatalogPackageMetadataExtraction.cs]: Keep consistency for different cases here.

File: src/Catalog/Registration/RegistrationMakerCatalogItem.cs [hunk 7]
```diff
@@ -161,22 +161,19 @@ private string GetLicenseUrl()
...
+            string packageId = _catalogItem.GetTriplesWithSubjectPredicate(subject, _catalogItem.CreateUriNode(Schema.Predicates.Id)).FirstOrDefault().Object.ToString();
+            string packageVersion = NuGetVersionUtility.NormalizeVersion(_catalogItem.GetTriplesWithSubjectPredicate(subject, _catalogItem.CreateUriNode(Schema.Predicates.Version)).FirstOrDefault().Object.ToString());
             Triple licenseExpression = _catalogItem.GetTriplesWithSubjectPredicate(subject, _catalogItem.CreateUriNode(Schema.Predicates.LicenseExpression)).FirstOrDefault();
             Triple licenseFile = _catalogItem.GetTriplesWithSubjectPredicate(subject, _catalogItem.CreateUriNode(Schema.Predicates.LicenseFile)).FirstOrDefault();
             Triple licenseUrl = _catalogItem.GetTriplesWithSubjectPredicate(subject, _catalogItem.CreateUriNode(Schema.Predicates.LicenseUrl)).FirstOrDefault();
 
-            if (_galleryBaseAddress != null)
+            if (_galleryBaseAddress != null &&
+                packageId != null &&
+                packageVersion != null &&
```
Comment (shishirx34) [inline src/Catalog/Registration/RegistrationMakerCatalogItem.cs]: for strings `string.IsNullOrEmpty` is more robust. I prefer using `string.IsNullOrWhitespace` which covers all string cases.

File: src/Catalog/Registration/RegistrationMakerCatalogItem.cs [hunk 8]
```diff
@@ -161,22 +161,19 @@ private string GetLicenseUrl()
...
+                ((licenseExpression != null && !string.IsNullOrWhiteSpace(licenseExpression.Object.ToString())) ||
+                 (licenseFile != null && !string.IsNullOrWhiteSpace(licenseFile.Object.ToString()))))
```
Comment (shishirx34) [inline src/Catalog/Registration/RegistrationMakerCatalogItem.cs]: you can merge this into single call like 
```cs
!string.IsNullOrWhiteSpace(licenseExpression?.Object.ToString()) || !string.IsNullOrWhiteSpace(licenseFile?.Object.ToString())
```

File: src/Catalog/Registration/RegistrationMakerCatalogItem.cs (hunk 7 above)
Comment (shishirx34) [reply src/Catalog/Registration/RegistrationMakerCatalogItem.cs]: do this for all strings above.

File: src/NuGet.Indexing/Extraction/CatalogPackageMetadataExtraction.cs [hunk 9]
```diff
@@ -55,18 +58,37 @@ private class Extractor
...
             }
 
+            private void AddLicenseUrl()
+            {
+                var packageId = JTokenToString(_catalog[MetadataConstants.IdPropertyName]);
+                var packageVersion = JTokenToString(_catalog[MetadataConstants.NormalizedVersionPropertyName]);
+
+                if (_galleryBaseAddress != null &&
+                    packageId != null &&
+                    packageVersion != null &&
```
Comment (shishirx34) [inline src/NuGet.Indexing/Extraction/CatalogPackageMetadataExtraction.cs]: why not use `!string.IsNullOrEmpty`?

File: src/Catalog/Helpers/LicenseHelper.cs [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+    {
+        /// <summary>
+        /// Generate the license url given package info and gallery base url.
+        /// </summary>
+        /// <param name="packageId">package Id</param>
+        /// <param name="packageVersion">package version</param>
+        /// <param name="galleryBaseAddress">url of gallery base address</param>
+        /// <returns>The url of license in gallery</returns>
+        public static string GetGalleryLicenseUrl(string packageId, string packageVersion, Uri galleryBaseAddress)
+        {
```
Comment (agr) [inline src/Catalog/Helpers/LicenseHelper.cs]: We should have `null` checks for public method arguments.

File: tests/CatalogTests/Helpers/LicenseHelperTests.cs [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace CatalogTests.Helpers
+{
+    public class LicenseHelperTests
+    {
+        [Theory]
+        [InlineData("testPackage", "1.0.0", "https://testnuget", "https://testnuget/packages/testPackage/1.0.0/license")]
+        [InlineData("testPackage", "1.0.0", "https://testnuget/", "https://testnuget/packages/testPackage/1.0.0/license")]
+        [InlineData("testPackage", "1.0.0", "https://testnuget//", "https://testnuget/packages/testPackage/1.0.0/license")]
+        public void GivenPackageIdAndVersionAndGalleryBaseUrl_ReturnsLicenseUrl(string packageId, string packageVersion, string galleryBaseAddress, string expectedLicenseUrl)
```
Comment (agr) [inline tests/CatalogTests/Helpers/LicenseHelperTests.cs]: I'd add a test that emphasizes the fact that no argument validation is done by the method besides `null` check. For example, pass some garbage as a version and check the result. Additionally, we need a test that checks the proper handling of characters outside of ASCII range.

Comment (agr) [reply tests/CatalogTests/Helpers/LicenseHelperTests.cs]: Example of such package: [测试更新包](https://www.nuget.org/packages/测试更新包/)

---
In reply to: [248849565](https://github.com/NuGet/NuGet.Services.Metadata/pull/470#discussion_r248849565) [](ancestors = 248849565)

File: src/NuGet.Indexing/NuGet.Indexing.csproj [hunk 12]
```diff
@@ -199,6 +199,15 @@
       <Version>8.2.1</Version>
     </PackageReference>
   </ItemGroup>
+  <ItemGroup>
+    <WCFMetadata Include="Connected Services\" />
+  </ItemGroup>
+  <ItemGroup>
```
Comment (agr) [inline src/NuGet.Indexing/NuGet.Indexing.csproj]: VS added it on its own, I presume?.. Should we keep it?
